File: mobile/ios/OpenU60/Features/Tools/SpeedTest/SpeedTestViewModel.swift
import SwiftUI

struct SpeedTestServer: Identifiable {
    let id: Int
    let name: String
    let country: String
    let sponsor: String
}

struct SpeedTestProgress {
    var phase: String = "idle"
    var progress: Int = 0
    var liveSpeedMbps: Double = 0
    var pingMs: Double?
    var jitterMs: Double?
    var downloadMbps: Double?
    var uploadMbps: Double?
    var downloadBytes: Int = 0
    var uploadBytes: Int = 0
    var server: String = ""
    var error: String?
}

@Observable
@MainActor
final class SpeedTestViewModel {
    var servers: [SpeedTestServer] = []
    var selectedServerId: Int?
    var progress: SpeedTestProgress = SpeedTestProgress()
    var isRunning: Bool = false
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    private let client: AgentClient
    private let authManager: AuthManager
    private nonisolated(unsafe) var pollTask: Task<Void, Never>?
    private var pollFailures = 0
    private let maxPollFailures = 10

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    deinit {
        pollTask?.cancel()
    }

    func loadServers() async {
        isLoading = true
        do {
            let items = try await client.getJSONArray("/api/speedtest/servers")
            servers = items.compactMap { dict -> SpeedTestServer? in
                guard let id = (dict["id"] as? Int) ?? (dict["id"] as? String).flatMap({ Int($0) }),
                      let name = dict["name"] as? String,
                      let country = dict["country"] as? String,
                      let sponsor = dict["sponsor"] as? String else { return nil }
                return SpeedTestServer(id: id, name: name, country: country, sponsor: sponsor)
            }
            if selectedServerId == nil, let first = servers.first {
                selectedServerId = first.id
            }
        } catch {
            showMessage("Failed to load servers: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func startTest() async {
        guard let serverId = selectedServerId else {
            showMessage("Select a server first", isError: true)
            return
        }

        isLoading = true
        message = nil
        progress = SpeedTestProgress()

        do {
            let _ = try await client.postJSON("/api/speedtest/start", body: ["server_id": serverId])
            isRunning = true
            showMessage("Speed test started", isError: false)
            startPolling()
        } catch {
            showMessage("Failed to start: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func stopTest() async {
        stopPolling()
        isRunning = false

        do {
            let _ = try await client.postJSON("/api/speedtest/stop")
            showMessage("Speed test stopped", isError: false)
        } catch {
            showMessage("Failed to stop: \(error.localizedDescription)", isError: true)
        }
    }

    private func startPolling() {
        stopPolling()
        pollFailures = 0
        pollTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(1))
                await pollProgress()
                if !isRunning { break }
            }
        }
    }

    private func stopPolling() {
        pollTask?.cancel()
        pollTask = nil
    }

    private func pollProgress() async {
        do {
            let data = try await client.getJSON("/api/speedtest/progress")
            pollFailures = 0

            progress.phase = data["phase"] as? String ?? progress.phase
            progress.progress = data["progress"] as? Int ?? progress.progress
            progress.liveSpeedMbps = data["live_speed_mbps"] as? Double ?? progress.liveSpeedMbps
            progress.pingMs = data["ping_ms"] as? Double
            progress.jitterMs = data["jitter_ms"] as? Double
            progress.downloadMbps = data["download_mbps"] as? Double
            progress.uploadMbps = data["upload_mbps"] as? Double
            progress.downloadBytes = data["download_bytes"] as? Int ?? progress.downloadBytes
            progress.uploadBytes = data["upload_bytes"] as? Int ?? progress.uploadBytes
            progress.server = data["server"] as? String ?? progress.server
            progress.error = data["error"] as? String

            if progress.phase == "complete" {
                isRunning = false
                stopPolling()
                showMessage("Speed test complete", isError: false)
            } else if progress.phase == "error" {
                isRunning = false
                stopPolling()
                showMessage(progress.error ?? "Speed test failed", isError: true)
            } else if progress.phase == "cancelled" {
                isRunning = false
                stopPolling()
                showMessage("Speed test cancelled", isError: false)
            }
        } catch {
            pollFailures += 1
            if pollFailures >= maxPollFailures {
                isRunning = false
                stopPolling()
                showMessage("Lost connection to device", isError: true)
            }
        }
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/tools/speedtest/SpeedTestViewModel.kt
package com.openu60.feature.tools.speedtest

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SpeedTestServerInfo(
    val id: Int,
    val name: String,
    val country: String,
    val sponsor: String,
)

data class SpeedTestState(
    val servers: List<SpeedTestServerInfo> = emptyList(),
    val selectedServerId: Int? = null,
    val phase: String = "idle",
    val progress: Int = 0,
    val liveSpeedMbps: Double = 0.0,
    val pingMs: Double? = null,
    val jitterMs: Double? = null,
    val downloadMbps: Double? = null,
    val uploadMbps: Double? = null,
    val downloadBytes: Long = 0,
    val uploadBytes: Long = 0,
    val serverName: String = "",
    val error: String? = null,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class SpeedTestViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(SpeedTestState())
    val state: StateFlow<SpeedTestState> = _state.asStateFlow()

    fun loadServers() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val list = agentClient.getJSONArray("/api/speedtest/servers")
                val servers = list.mapNotNull { map ->
                    val id = (map["id"] as? Number)?.toInt() ?: return@mapNotNull null
                    val name = map["name"] as? String ?: ""
                    val country = map["country"] as? String ?: ""
                    val sponsor = map["sponsor"] as? String ?: ""
                    SpeedTestServerInfo(id, name, country, sponsor)
                }
                _state.value = _state.value.copy(
                    servers = servers,
                    selectedServerId = _state.value.selectedServerId ?: servers.firstOrNull()?.id,
                    isLoading = false,
                )
            } catch (e: AgentError.Unauthorized) {
                authManager.reauthenticate()
                setError("Session expired — please try again")
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun selectServer(id: Int) {
        _state.value = _state.value.copy(selectedServerId = id)
    }

    fun startTest() {
        val serverId = _state.value.selectedServerId ?: return
        viewModelScope.launch {
            _state.value = _state.value.copy(
                isLoading = true,
                message = null,
                progress = 0,
                liveSpeedMbps = 0.0,
                pingMs = null,
                jitterMs = null,
                downloadMbps = null,
                uploadMbps = null,
                downloadBytes = 0,
                uploadBytes = 0,
                error = null,
            )
            try {
                agentClient.postJSON("/api/speedtest/start", mapOf("server_id" to serverId))
                _state.value = _state.value.copy(isLoading = false)
                // Poll until done (max 5 minutes)
                var pollCount = 0
                while (pollCount < 300) {
                    pollCount++
                    delay(1000)
                    val data = agentClient.getJSON("/api/speedtest/progress")
                    val phase = data["phase"] as? String ?: "idle"
                    val progress = (data["progress"] as? Number)?.toInt() ?: 0
                    val liveSpeed = (data["live_speed_mbps"] as? Number)?.toDouble() ?: 0.0
                    val ping = (data["ping_ms"] as? Number)?.toDouble()
                    val jitter = (data["jitter_ms"] as? Number)?.toDouble()
                    val download = (data["download_mbps"] as? Number)?.toDouble()
                    val upload = (data["upload_mbps"] as? Number)?.toDouble()
                    val dlBytes = (data["download_bytes"] as? Number)?.toLong() ?: 0L
                    val ulBytes = (data["upload_bytes"] as? Number)?.toLong() ?: 0L
                    val serverName = data["server"] as? String ?: ""
                    val errorMsg = data["error"] as? String

                    _state.value = _state.value.copy(
                        phase = phase,
                        progress = progress,
                        liveSpeedMbps = liveSpeed,
                        pingMs = ping,
                        jitterMs = jitter,
                        downloadMbps = download,
                        uploadMbps = upload,
                        downloadBytes = dlBytes,
                        uploadBytes = ulBytes,
                        serverName = serverName,
                        error = errorMsg,
                    )

                    if (phase in listOf("complete", "cancelled", "error")) break
                }
                if (pollCount >= 300) {
                    setError("Speed test timed out")
                    return@launch
                }
                val finalPhase = _state.value.phase
                _state.value = _state.value.copy(
                    isLoading = false,
                    message = when (finalPhase) {
                        "complete" -> "Speed test complete"
                        "cancelled" -> "Speed test cancelled"
                        "error" -> _state.value.error ?: "Speed test failed"
                        else -> null
                    },
                    messageIsError = finalPhase == "error",
                )
            } catch (e: AgentError.Unauthorized) {
                authManager.reauthenticate()
                setError("Session expired — please try again")
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun stopTest() {
        viewModelScope.launch {
            try {
                agentClient.postJSON("/api/speedtest/stop")
            } catch (e: AgentError.Unauthorized) {
                authManager.reauthenticate()
                setError("Session expired — please try again")
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(
            isLoading = false,
            phase = "idle",
            message = msg ?: "Unknown error",
            messageIsError = true,
        )
    }
}
